frontend/notification/pushNotificationService.js

```diff
@@ -0,0 +1,130 @@
+
+        service.currentUser = AuthService.getCurrentUser();
+
+        /**
+         * Ask permission to the user to send push notifications
+         * and if the permission is conceded the user's new token
+         * is retrieved and saveToken is called passing the token
+         * as parameter.
+         * @private
+         */
+        service._requestNotificationPermission = function requestNotificationPermission() {
+            messaging.requestPermission().then(() => {
+                messaging.getToken().then(token => {
+                    service._saveToken(token);
+                }, () => {
+                    MessageService.showToast('Não foi possível ativar as notificações.');
+                });
+            });
+        };
+
+        /**
+         * It receives a token, starts a reference to
+         * firebase's database and save the token.
+         * @param {String} token
+         * @private 
+         */
+        service._saveToken = function saveToken(token) {
+            const notificationsRef = service._initFirebaseArray();
+            service._setToken(token, notificationsRef);
+        };
+
+        /**
+         * Instantiate a reference to the database
+         * based on the userKey, starts a firebaseArray
+         * in case of it hasn't been started before, and
+         * return the reference.
+         * @private
+         */
+        service._initFirebaseArray = function initFirebaseArray() {
+            const endPoint = `${PUSH_NOTIFICATIONS_URL}${service.currentUser.key}`;
+            const notificationsRef = ref.child(endPoint);
+
+            if (!service.firebaseArrayNotifications) {
+                service.firebaseArrayNotifications = $firebaseArray(notificationsRef);
+            }
+
+            return notificationsRef;
+        };
+
+        /**
+         * It is responsible for check if the user already have a token.
+         * If it does, the token is replaced by the new one received as parameter.
+         * Otherwise the token is saved.
+         * @param {String} token 
+         * @param {object} notificationsRef
+         * @private 
+         */
+        service._setToken = function setToken(token, notificationsRef) {
+            service.firebaseArrayNotifications.$loaded().then(() => {
+                const tokenObject = $firebaseObject(notificationsRef);
+                tokenObject.token = token;
+                service.firebaseArrayNotifications.$add(tokenObject);
+            });
+        };
+
+        /**
+         * Check if the user has already conceded the permission
+         * using Notification object.
+         * @private
+         */
+        service._hasNotificationPermission = function hasNotificationPermission() {
+            const { permission } = Notification;
+            return permission === "granted";
+        };
+
+        (function init() {
+            if (!service._hasNotificationPermission() && isMobile.any()) {
```

Comment:
If the permission should only be requested as soon as the user logs it does not make sense to place the request in the init of the service.